### PR TITLE
HS-1332: Passive discovery validations

### PR DESCRIPTION
**Pull request by rabzobi**
## Description
Allow update to keep the same location
Validate community string according to rules in ticket

## Jira link(s)
- https://issues.opennms.org/browse/HS-1332

## Flagged for review
Not sure of validation requirement for community strings and comma's. I have implemented as per my understanding from the ticket.

## Checklist
* [X] Follows Horizon Stream's [development guidelines.](https://github.com/OpenNMS/horizon-stream/wiki/Development-Guidelines)
* [X] Appropriate reviewer(s) have been selected.
* [X] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS/horizon-stream/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts

### CODE REVIEW

**Review by james8hut — CHANGES_REQUESTED** [review]
Any JUnit tests?

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/discovery/PassiveDiscoveryService.java [hunk 1]
```diff
@@ -160,6 +163,22 @@ private void validateSnmpPorts(PassiveDiscoveryUpsertDTO dto) {
         }
     }
 
+    private void validateCommunityStrings(PassiveDiscoveryUpsertDTO passiveDiscovery) {
+        String snmpCommunities = "";
+        for (String snmpCommunity: passiveDiscovery.getCommunitiesList()){
+            snmpCommunities += snmpCommunity.replace(",","") + " "; // As be requirement on HS-1332
```
Comment (james8hut) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/discovery/PassiveDiscoveryService.java:169]: I'm not sure what the comment means exactly.
I'd also be inclined to remove it anyway.

Comment (rabzobi) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/discovery/PassiveDiscoveryService.java]: It was supposed to say "as per"... but I have removed it

**Comment by rabzobi** [comment]
> Any JUnit tests?

I have added some tests

File: inventory/main/src/test/java/org/opennms/horizon/inventory/service/PassiveDiscoveryServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+        Exception exception = assertThrows(InventoryRuntimeException.class, () -> {
+            List<String> communities = new ArrayList<>();
+            communities.add("123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789");
+            PassiveDiscoveryUpsertDTO tooLong = PassiveDiscoveryUpsertDTO
+                .newBuilder().addAllCommunities(communities)
+                .build();
+            passiveDiscoveryService.validateCommunityStrings(tooLong);
+        });
+        assertTrue(exception.getMessage().equals("Snmp communities string is too long"));
```
Comment (james8hut) [inline inventory/main/src/test/java/org/opennms/horizon/inventory/service/PassiveDiscoveryServiceTest.java:80]: I would create a separate test method for each of these things. If the too long test fails then the 7 bit ascii test doesn't get to run, so I always find it annoying as a developer, if I was to fix the too long test, and only then find out the 7 bit ascii test was going to fail as well.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/discovery/PassiveDiscoveryService.java [hunk 3]
```diff
@@ -97,8 +97,11 @@ public PassiveDiscoveryDTO updateDiscovery(String tenantId, PassiveDiscoveryUpse
             throw new InventoryRuntimeException("Passive discovery not found for id: " + id);
         }
 
-        validateDiscovery(tenantId, request);
+        if (!request.getLocation().equals(discoveryOpt.get().getLocation())) {
+            validateDiscovery(tenantId, request);
```
Comment (james8hut) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/discovery/PassiveDiscoveryService.java:101]: Can you add a test to ensure validateDiscovery method isn't called.
Plan to do this:
Comment out the if statement
Test should validateDiscovery and throw an exception. (And fail)
Uncomment the if statement
Test now should pass, as it doesn't call validateDiscovery.

**Review by james8hut — CHANGES_REQUESTED** [review]

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/discovery/PassiveDiscoveryService.java (hunk 3 above)
Comment (rabzobi) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/discovery/PassiveDiscoveryService.java:101]: I've added two heavily mocked test cases which cover the exception scenario (and no exception scenario)

File: inventory/main/src/test/java/org/opennms/horizon/inventory/service/PassiveDiscoveryServiceTest.java (hunk 2 above)
Comment (rabzobi) [reply inventory/main/src/test/java/org/opennms/horizon/inventory/service/PassiveDiscoveryServiceTest.java]: I have moved them all into separate test methods

**Review by james8hut — APPROVED** [review]

**Pull request merged**